src/lib.rs

```diff
@@ -203,6 +206,41 @@ impl CodecPipelineImpl {
     }
 }
 
+fn array_metadata_to_codec_metadata_v3(
+    metadata: ArrayMetadata,
+) -> Result<Vec<MetadataV3>, ArrayMetadataV2ToV3Error> {
+    match metadata {
+        ArrayMetadata::V3(metadata) => Ok(metadata.codecs),
+        ArrayMetadata::V2(metadata) => {
+            let config = global_config();
+            let (Ok(data_type), endianness) = (
+                data_type_metadata_v2_to_v3(
+                    &metadata.dtype,
+                    config.data_type_aliases_v2(),
+                    config.data_type_aliases_v3(),
+                ),
+                data_type_metadata_v2_to_endianness(&metadata.dtype)
+                    .map_err(ArrayMetadataV2ToV3Error::InvalidEndianness)?,
+            ) else {
+                return Err(ArrayMetadataV2ToV3Error::UnsupportedDataType(
+                    metadata.dtype.clone(),
+                ));
+            };
```

Comment:
This is a bit convoluted, it does:
1. run `data_type_metadata_v2_to_endianness` and short-circuit the error using `?`
2. run `data_type_metadata_v2_to_v3` and map the error to a `UnsupportedDataType` error

And the tuple destructuring that happens doesn’t seem to serve any purpose.

Why not just

```rust
let endianness = data_type_metadata_v2_to_endianness(...).map_err(...)?;
let data_type = data_type_metadata_v2_to_v3(...).map_err(...)?;
```

it does the same and is much cleaner, no?

Reply:
I took the liberty to change this, if execution order matters, we’re in trouble anyway.

Reply:
Nice one, much cleaner